ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +374,186 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
+          nonIteratingNonPendingRemoteReplicas.remove(remoteReplicaInfo);
+          throttleCount.inc();
+        }
+      });
+
+      if(inflightRemoteReplicaGroup.containsKey(groupId)){
+        nonIteratingNonPendingRemoteReplicas.removeAll(inflightRemoteReplicaGroup.get(groupId).remoteReplicaInfos);
+      }
+
+      if(inflightRemoteReplicaGroup.containsKey(standbyReplicasNoProgressReplicaGroupId)) {
+        nonIteratingNonPendingRemoteReplicas.removeAll(inflightRemoteReplicaGroup.get(standbyReplicasNoProgressReplicaGroupId).remoteReplicaInfos);
+      }
+
+      List<RemoteReplicaInfo> activeReplicas = new ArrayList<>();
+      List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+
+      filterRemoteReplicasToReplicate(new ArrayList<>(nonIteratingNonPendingRemoteReplicas), activeReplicas, standbyReplicasWithNoProgress);
+
+      if (!inflightRemoteReplicaGroup.containsKey(groupId)) {
+        if (!activeReplicas.isEmpty()) {
+          RemoteReplicaGroup groupForActiveReplicas = new RemoteReplicaGroup(activeReplicas, remoteNode, false, groupId);
+          inflightRemoteReplicaGroup.put(groupId, groupForActiveReplicas);
+          groupIdIterationCountMap.put(groupId, groupIdIterationCountMap.getOrDefault(groupId, 0) + 1);
+        }
+      }
+
+      List<RemoteReplicaInfo> standByReplicasTimedOutOnNoProgress =
+          getRemoteStandbyReplicasTimedOutOnNoProgress(standbyReplicasWithNoProgress);
+
+      if (standByReplicasTimedOutOnNoProgress.isEmpty()) {
+        continue;
+      }
+
+      dataNodeIdToPendingReplicasMarkedForStandByNoProgress.putIfAbsent(remoteNode, new HashSet<>());
+      dataNodeIdToPendingReplicasMarkedForStandByNoProgress.get(remoteNode).addAll(standByReplicasTimedOutOnNoProgress);
+    }
+
+    for(Map.Entry<DataNodeId, Set<RemoteReplicaInfo>> entry : dataNodeIdToPendingReplicasMarkedForStandByNoProgress.entrySet()){
+      DataNodeId remoteNode = entry.getKey();
+      Integer groupId = dataNodeIdStandbyReplicasNoProgressReplicaGroupId.get(remoteNode);
+
+      if(inflightRemoteReplicaGroup.containsKey(groupId))
+        continue;
+
+
+      List<RemoteReplicaInfo> remoteReplicaInfos = new ArrayList<>(entry.getValue());
+      if(remoteReplicaInfos.isEmpty())
+        continue;
+
+      RemoteReplicaGroup remoteReplicaGroup = new RemoteReplicaGroup(remoteReplicaInfos, remoteNode, true, groupId);
+      inflightRemoteReplicaGroup.put(groupId, remoteReplicaGroup);
+      groupIdIterationCountMap.put(groupId, groupIdIterationCountMap.getOrDefault(groupId, 0) + 1);
+
+      dataNodeIdToPendingReplicasMarkedForStandByNoProgress.get(remoteNode).clear();
+    }
+
+    if (inflightRemoteReplicaGroup.isEmpty()) {
+      allReplicasCaughtUpEarly.setTrue();
+    }
+  }
+
+  /**
+   * Checks if acyclic replication is enabled and calls appropriate logic for replication
+   */
+  public void replicate(){
+    if(enableAcyclicReplication)
+      replicateAcyclic();
```

Comment:
its not truly acyclic right ? the name is confusing.

Reply:
changed the name to replicateLimitedAcyclic